### PR TITLE
8289711: Add container configuration data to mbeans

### PR DESCRIPTION
**Pull request by xpbob**
Container configuration information is useful for troubleshooting problems,Exposing information in MBeans is ideal for monitoring, jConsole, and other scenarios.
Results the following
图片

---------
### Progress
- [ ] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [ ] Change requires a CSR request to be approved

### Issue
 * [JDK-8289711](https://bugs.openjdk.org/browse/JDK-8289711): Add container configuration data to mbeans

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/9372/head:pull/9372` \
`$ git checkout pull/9372`

Update a local copy of the PR: \
`$ git checkout pull/9372` \
`$ git pull https://git.openjdk.org/jdk pull/9372/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 9372`

View PR using the GUI difftool: \
`$ git pr show -t 9372`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/9372.diff">https://git.openjdk.org/jdk/pull/9372.diff</a>

</details>

### CODE REVIEW

File: src/java.base/share/classes/module-info.java [hunk 1]
```diff
@@ -230,6 +230,7 @@
         jdk.internal.jvmstat;
     exports jdk.internal.platform to
         jdk.management,
+        java.management,
```
Comment (DamonFool) [inline src/java.base/share/classes/module-info.java:233]: Maybe, we'd better put `jdk.management,` before `jdk.management,`.

File: src/java.management/share/classes/java/lang/management/ContainerMXBean.java [hunk 2]
```diff
@@ -0,0 +1,138 @@
+package java.lang.management;
```
Comment (DamonFool) [inline src/java.management/share/classes/java/lang/management/ContainerMXBean.java:1]: Shall we add a copyright header for this new file?

**Comment by xpbob** [comment]
@DamonFool 
Thanks for review.
The code has been updated

**Comment by AlanBateman** [comment]
It's not clear that introducing this as a standard API is the right thing to do. Are you 100% confident that the concepts of "CPU quota", "CPU shares", "CPU period", "soft limit" etc. will last the test of time and that we don't be back here next year with another PR to deprecate or replace this API?  I don't disagree that exposing a MXBean could be useful for monitoring/management purposes but I think we have to be cautious getting locked in. A possible standard point for prototyping purposes and getting feedback is a JDK-specific MXBean (module jdk.management).

**Comment by jerboaa** [comment]
I do think this could be useful.

Having said that, some of this has been discussed in https://bugs.openjdk.org/browse/JDK-8199944 (and been considered as won't fix at the time). I tend to agree that it should be a JDK (and even an OS) specific bean. The relevant information is only available on Linux. +1 on making this JDK specific. First step is probably to get consensus on a CSR.

**Comment by xpbob** [comment]
@AlanBateman @jerboaa That's a good idea.Adding a special bean is only available on Linux systems.I do not know the process of creating a CSR, can you help me create a CSR

**Comment by jerboaa** [comment]
> @AlanBateman @jerboaa That's a good idea.Adding a special bean is only available on Linux systems.I do not know the process of creating a CSR, can you help me create a CSR

See https://wiki.openjdk.org/display/csr/CSR+FAQs

**Comment by poonamparhar** [comment]
@AlanBateman OperatingSystemMXBean makes the CPU and Memory related information for a process available regardless of whether that process is running in a native operating system environment or a containerized environment. It does not provide container configuration information such as its provider (crgoups v1 or v2), CPU shares, CPU quota etc, which can be useful for monitoring and troubleshooting purposes.

**Comment by AlanBateman** [comment]
> @AlanBateman OperatingSystemMXBean makes the CPU and Memory related information for a process available regardless of whether that process is running in a native operating system environment or a containerized environment. It does not provide container configuration information such as its provider (crgoups v1 or v2), CPU shares, CPU quota etc, which can be useful for monitoring and troubleshooting purposes.

Yes, I know this but I think there is a lot more thought required before deciding to augment this with another API for the container environment. There are a couple of things to explore.

**Comment by xpbob** [comment]
@AlanBateman @jerboaa @poonamparhar @DamonFool 
Thanks for review.
I add mBeans using the registerMBean method.
We can get configuration information through JConsole, JMX exporter

**Comment by AlanBateman** [comment]
> Thanks for review.
> I add mBeans using the registerMBean method.
> We can get configuration information through JConsole, JMX exporter

This iteration is a bit confusing because it adds a public interface to java.lang.management. For the registerMBean prototype then you shouldn't need any API changes to the java.management module.

**Comment by xpbob** [comment]
> > Thanks for review.
> > I add mBeans using the registerMBean method.
> > We can get configuration information through JConsole, JMX exporter
> 
> This iteration is a bit confusing because it adds a public interface to java.lang.management. For the registerMBean prototype then you shouldn't need any API changes to the java.management module.
@AlanBateman Thanks for review.
The runtime information is already fetched through the OperatingSystemMXBean, and I'm wrapped through the interface, keeping only the configuration data.Such an interface could be part of management.

**Comment by xpbob** [comment]
Thanks for review. @AlanBateman 
I move ContainerInfoMXBean to sun.management, the data is not available,
图片
The red character indicates that it is not available

File: src/java.management/share/classes/module-info.java [hunk 3]
```diff
@@ -49,6 +49,7 @@
     exports javax.management.relation;
     exports javax.management.remote;
     exports javax.management.timer;
+    exports java.lang.management.internal.mBeans;
```
Comment (AlanBateman) [inline src/java.management/share/classes/module-info.java:52]: This package should not be exported (at least not to all modules).

**Comment by AlanBateman** [comment]
What happened with the experiment to move ContainerInfoMXBean to jdk.management?

**Comment by xpbob** [comment]
Thanks for review.
@AlanBateman 
I found my code bug and it has been fixed.
Moving ContainerInfo is feasible

**Comment by xpbob** [comment]
图片

**Comment by xpbob** [comment]
Hi, @AlanBateman 
I updated the code and move ContainerInfoMXBean to jdk.management
Can the CSR tag be removed?

File: src/jdk.management/share/classes/com/sun/management/ContainerInfoMXBean.java [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+ */
+
+package com.sun.management;
+
+import java.lang.management.PlatformManagedObject;
+
+/**
+ * This is a special bean , only available on Linux systems
+ */
+public interface ContainerInfoMXBean  extends PlatformManagedObject {
```
Comment (turbanoff) [inline src/jdk.management/share/classes/com/sun/management/ContainerInfoMXBean.java:31]: let's remove one redundant space before `extends`

**Pull request closed without merging**